### PR TITLE
[feature](nereids)support bitmap runtime filter on nereids

### LINKED ISSUE #17506: support bitmap type as join's hash conjuncts type in Nereids

### PR DESCRIPTION
**Pull request by englefly**
# Proposed changes

`select a from t where a in (select k from bitmap_table)`
, in which `t.a` is int, and `bitmap_table.k` is bitmap
this sql is transformed to 
`select a from t left semi join (select bitmap_union(k2) u from bitmap_table) t2 on bitmap_contains(t2.u, t.a)`
and then we generate runtime filter for expression `bitmap_contains(t2.u, t.a)`

original planner only support in-subquery

Issue Number: close #17506

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.76 seconds
 stream load tsv:          465 seconds loaded 74807831229 Bytes, about 153 MB/s
 stream load json:         37 seconds loaded 2358488459 Bytes, about 60 MB/s
 stream load orc:          67 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230220080658_clickbench_pr_100038.html

**Comment by morrySnow** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/RuntimeFilter.java [hunk 1]
```diff
@@ -30,21 +30,37 @@ public class RuntimeFilter {
...
     private final Expression srcSlot;
+    //bitmap filter support target expression like  k1+1, abs(k1)
+    //targetExpression is an expression on targetSlot, in which there is only one non-const slot
+    private Expression targetExpression;
     private Slot targetSlot;
     private final int exprOrder;
-    private PhysicalHashJoin builderNode;
+    private AbstractPhysicalJoin builderNode;
+
+    private boolean bitmapFilterNotIn;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/RuntimeFilter.java:40]: the attr name could be more generation such as `reverse` or `anti`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 2]
```diff
@@ -136,6 +139,44 @@ public static Pair<List<Expression>, List<Expression>> extractExpressionForHashT
...
+                bitmapContains = (BitmapContains) expr;
+            }
+            if (bitmapContains == null) {
+                continue;
+            }
+            //first child in right, second child in left
+            if (leftSlots.containsAll(bitmapContains.child(1).collect(Slot.class::isInstance))
+                    && rightSlots.containsAll(bitmapContains.child(0).collect(Slot.class::isInstance))) {
+                result.add(expr);
+            }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:175]: add a `left contains child(0) && right contains child(1)` branch is more robustness

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 3]
```diff
@@ -136,6 +139,44 @@ public static Pair<List<Expression>, List<Expression>> extractExpressionForHashT
         );
     }
 
+    public static boolean hasBitmapSlot(Expression expr) {
+        return expr.getInputSlots().stream()
+                .anyMatch(slot -> slot.getDataType() == BitmapType.INSTANCE);
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:145]: not use any more

File: regression-test/suites/query_p0/join/test_bitmap_filter_nereids.groovy [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    sql """
+    insert into ${tbl2} values 
+    (1, bitmap_from_string('1, 3, 5, 7, 9, 11, 13, 99, 19910811, 20150402'),
+        bitmap_from_string('32767, 1985, 255, 789, 1991')),
+    (2, bitmap_from_string('10, 11, 12, 13, 14'), bitmap_empty());"""
+
+    sql "set enable_nereids_planner=true;"
+    sql "set enable_fallback_to_original_planner=false;"
+
+    qt_sql1 "select k1, k2 from ${tbl1} where k1 in (select k2 from ${tbl2}) order by k1;"
```
Comment (morrySnow) [inline regression-test/suites/query_p0/join/test_bitmap_filter_nereids.groovy:48]: add correlation subquery cases

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 5]
```diff
@@ -1170,15 +1184,17 @@ public PlanFragment visitPhysicalNestedLoopJoin(
...
-            if (needNewRootFragment) {
-                connectChildFragment(nestedLoopJoinNode, 0, joinFragment, leftFragment, context);
-            } else {
-                nestedLoopJoinNode.setChild(0, leftFragment.getPlanRoot());
-                joinFragment.setPlanRoot(nestedLoopJoinNode);
-            }
+
             connectChildFragment(nestedLoopJoinNode, 1, joinFragment, rightFragment, context);
             List<Expr> joinConjuncts = nestedLoopJoin.getOtherJoinConjuncts().stream()
+                    .filter(e -> !nestedLoopJoin.isBitmapRuntimeFilterCondition(e))
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:1190]: why add this, as we discussion, we do not need to remove any conjunct?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommute.java [hunk 6]
```diff
@@ -93,4 +99,21 @@ private static boolean containJoin(GroupPlan groupPlan) {
         List<Slot> output = groupPlan.getOutput();
         return !output.stream().map(Slot::getQualifier).allMatch(output.get(0).getQualifier()::equals);
     }
+
+    private boolean disableCommunicateForBitmapRuntimeFilter(LogicalJoin<GroupPlan, GroupPlan> join) {
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommute.java:103]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java [hunk 7]
```diff
@@ -101,6 +101,7 @@ public enum RuleType {
     INFER_FILTER_NOT_NULL(RuleTypeClass.REWRITE),
     INFER_JOIN_NOT_NULL(RuleTypeClass.REWRITE),
     // subquery analyze
+    ANALYZE_FILTER_SUBQUERY_REWRITE_BITMAP_IN(RuleTypeClass.REWRITE),
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java:104]: removed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/processor/post/RuntimeFilterGenerator.java [hunk 8]
```diff
@@ -94,6 +97,9 @@ public PhysicalPlan visitPhysicalHashJoin(PhysicalHashJoin<? extends Plan, ? ext
                 EqualTo equalTo = ((EqualTo) JoinUtils.swapEqualToForChildrenOrder(
                         (EqualTo) join.getHashJoinConjuncts().get(i), join.left().getOutputSet()));
                 for (TRuntimeFilterType type : legalTypes) {
+                    if (type == TRuntimeFilterType.BITMAP) {
+                        continue;
+                    }
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/processor/post/RuntimeFilterGenerator.java:102]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalFilter.java [hunk 9]
```diff
@@ -150,4 +150,9 @@ public boolean isSingleTableExpressionExtracted() {
         return singleTableExpressionExtracted;
     }
 
+    public Plan withConjuncts(Set<Expression> conjuncts) {
+        return new LogicalFilter(conjuncts, this.groupExpression,
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalFilter.java:154]: unused code. removed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java [hunk 10]
```diff
@@ -45,22 +52,52 @@ public class InApplyToJoin extends OneRewriteRuleFactory {
...
+                }
+                return new LogicalJoin<>(JoinType.LEFT_SEMI_JOIN, Lists.newArrayList(),
+                        Lists.newArrayList(expr),
+                        JoinHint.NONE,
+                        apply.left(), agg);
+            }
+
+            //in-predicate to equal
+            Expression predicate = new EqualTo(((InSubquery) apply.getSubqueryExpr()).getCompareExpr(),
+                    apply.right().getOutput().get(0));
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java:88]: good for read.
In original code, the else-body is also in if-block

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java [hunk 11]
```diff
@@ -45,22 +52,52 @@ public class InApplyToJoin extends OneRewriteRuleFactory {
...
-                predicate = new EqualTo(((InSubquery) apply.getSubqueryExpr()).getCompareExpr(),
-                        apply.right().getOutput().get(0));
             }
-
-            //TODO nereids should support bitmap runtime filter in future
             List<Expression> conjuncts = ExpressionUtils.extractConjunction(predicate);
-            if (conjuncts.stream().anyMatch(expression -> expression.children().stream()
-                    .anyMatch(expr -> expr.getDataType() == BitmapType.INSTANCE))) {
-                throw new AnalysisException("nereids don't support bitmap runtime filter");
+            if (predicate instanceof BitmapContains) {
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java:96]: BitmapContains is generated in function: convertInPredicateToJoinConjunct() line 129
this "if" cannot be removed. we transform  " k in (select A ...)" to a join.
there are two types of in
1. A is comparable type to k, for example, A is int, and k is int. in-predicate becomes to "k=A"
2. A is bitmap, k is not. for example, k is int, A is bitmap. in-predicate becomes "bitmap_contains(A, k)"

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java [hunk 12]
```diff
@@ -77,4 +114,22 @@ public Rule build() {
             }
         }).toRule(RuleType.IN_APPLY_TO_JOIN);
     }
+
+    private boolean needBitmapUnoin(LogicalApply<GroupPlan, GroupPlan> apply) {
+        return apply.right().getOutput().get(0).getDataType().isBitmapType()
+                && !((InSubquery) apply.getSubqueryExpr()).getCompareExpr().getDataType().isBitmapType();
+    }
+
+    private Expression convertInPredicateToJoinConjunct(LogicalApply<GroupPlan, GroupPlan> apply) {
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java:123]: yes

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java [hunk 13]
```diff
@@ -45,22 +52,52 @@ public class InApplyToJoin extends OneRewriteRuleFactory {
...
+            if (needBitmapUnoin(apply)) {
+                /*
+                select t1.k1 from bigtable t1 where t1.k1 in (select t2.k2 from bitmap_table t2);
+                =>
+                select t1.k1 from bigtable t1 where t1.k1 in (select bitmap_union(k2) from bitmap_table t2);
+                =>
+                select t1.k1 from bigtable t1 left semi join (select bitmap_union(k2) x from bitmap_table ) t2
+                on bitmap_contains(x, t1.k1);
+                 */
+                List<Expression> groupExpressions = Lists.newArrayList();
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java:64]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java [hunk 14]
```diff
@@ -77,4 +114,22 @@ public Rule build() {
             }
         }).toRule(RuleType.IN_APPLY_TO_JOIN);
     }
+
+    private boolean needBitmapUnoin(LogicalApply<GroupPlan, GroupPlan> apply) {
+        return apply.right().getOutput().get(0).getDataType().isBitmapType()
+                && !((InSubquery) apply.getSubqueryExpr()).getCompareExpr().getDataType().isBitmapType();
+    }
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java:121]: yes. bitmap_union also works for "in" and no extra cost

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalNestedLoopJoin.java [hunk 15]
```diff
@@ -42,6 +44,13 @@
         RIGHT_CHILD_TYPE extends Plan>
         extends AbstractPhysicalJoin<LEFT_CHILD_TYPE, RIGHT_CHILD_TYPE> {
 
+    /*
+    bitmap_contains(...) or Not(bitmap_contains(...)) can be used as bitmap runtime filter condition
+    bitmapRF is different from other RF in that scan node must wait for it.
+    if a condition is used in rf, it can be removed from join conditions. we collect these conditions here.
+     */
+    private Set<Expression> bitMapRuntimeFilterConditions = Sets.newHashSet();
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalNestedLoopJoin.java:52]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java [hunk 16]
```diff
@@ -45,22 +52,52 @@ public class InApplyToJoin extends OneRewriteRuleFactory {
...
         return logicalApply().when(LogicalApply::isIn).then(apply -> {
-            Expression predicate;
+            if (needBitmapUnoin(apply)) {
+                /*
+                select t1.k1 from bigtable t1 where t1.k1 in (select t2.k2 from bitmap_table t2);
+                =>
+                select t1.k1 from bigtable t1 where t1.k1 in (select bitmap_union(k2) from bitmap_table t2);
+                =>
+                select t1.k1 from bigtable t1 left semi join (select bitmap_union(k2) x from bitmap_table ) t2
+                on bitmap_contains(x, t1.k1);
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InApplyToJoin.java:62]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/RuntimeFilter.java (hunk 1 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/RuntimeFilter.java:40]: as we discussed, BE only support semi, not anti.

File: regression-test/suites/query_p0/join/test_bitmap_filter_nereids.groovy (hunk 4 above)
Comment (englefly) [reply regression-test/suites/query_p0/join/test_bitmap_filter_nereids.groovy:48]: correlation subquery is not supported by "in (select bitmap...)"

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java (hunk 3 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:145]: removed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java (hunk 5 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:1190]: if the conjunct is used in bitmap rf, it is always true here

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Review by starocean999 — APPROVED** [review]
LGTM

**Pull request merged**